redis_failover backed by ZooKeeper

ryanlecompte opened the pull request.

ryanlecompte committed:
initial cut of zookeeper integration

ryanlecompte committed:
use the more actively maintained zk gem

ryanlecompte committed:
cleanup

ryanlecompte committed:
logging, handle startup edge case

ryanlecompte committed:
wording rename

ryanlecompte committed:
better exception handling for ZK

ryanlecompte committed:
readme updates for zookeeper

ryanlecompte committed:
readme tweaks

ryanlecompte committed:
re-watch as soon as possible

ryanlecompte committed:
more README updates

ryanlecompte committed:
more README tweaks

ryanlecompte committed:
tweaks to README

ryanlecompte commented:
Note that the main thing we are trying to address here with now relying on ZooKeeper is to solve network partition issues that plague the current implementation. Here's a description of the issue by @freels on the sf-ruby mailing list:

"Consider a scenario where I have some clients, two redis nodes, and one failover server: The failover server starts intermittently flapping. Maybe it's swapping, maybe there's some network misconfiguration, or some bizarre hardware failure, doesn't matter. From the perspective of clients, the failover server is popping in and out of existence. However, from the perspective of the failover server, the _rest_ of the system is popping in and out of existence, including the master node, so it starts to advertise a new master. Now all this time, clients are checking in with the failover server to update their node list. Some of them receive the new master, but some of them timeout since the failover server is still flapping. The ones that timeout continue to use the original server list. At this point, you're done, since you have two sets of clients talking to different masters."

ryanlecompte committed:
new gem description/summary

ryanlecompte committed:
add ZK client wrapper to deal with session expirations

ryanlecompte committed:
fix specs

ryanlecompte committed:
style change

ryanlecompte committed:
exception cleanup

ryanlecompte committed:
use reopen

ryanlecompte commented:
One concern that I have is what I just read in the ZooKeeper docs, which makes me think that there could still be cases where two different clients could end up writing to two different masters:

Sometimes developers mistakenly assume one other guarantee that ZooKeeper does not in fact make. This is:

Simultaneously Consistent Cross-Client Views
ZooKeeper does not guarantee that at every instance in time, two different clients will have identical views of ZooKeeper data. Due to factors like network delays, one client may perform an update before another client gets notified of the change. Consider the scenario of two clients, A and B. If client A sets the value of a znode /a from 0 to 1, then tells client B to read /a, client B may read the old value of 0, depending on which server it is connected to. If it is important that Client A and Client B read the same value, Client B should should call the sync() method from the ZooKeeper API method before it performs its read.

So, ZooKeeper by itself doesn't guarantee that changes occur synchronously across all servers, but ZooKeeper primitives can be used to construct higher level functions that provide useful client synchronization. (For more information, see the ZooKeeper Recipes. [tbd:..]).

(from http://zookeeper.apache.org/doc/trunk/zookeeperProgrammers.html)

eric commented on lib/redis_failover/client.rb:
```diff
@@ -159,39 +155,43 @@ def slave
-      end
+      data = @zkclient.get(ZK_PATH, :watch => true).first
+      # register a watcher for future changes
+      @zkclient.watcher.register(ZK_PATH) { build_clients }
```
I believe you only need to register once — I'm not sure what happens if you do it more than once.

Also, I believe (but you'll have to check on it), that you'll get lifecycle events (connected, disconnected, etc) along with events about the node you're watching changing, so it may be worth actually checking what the event you received was.

ryanlecompte replied:
I was actually re-registering because the ZK docs say that a watch is only fired once. If you want to keep getting watch events, you need to re-register your watch after reading. Maybe the ZK gem does this for us behind the scenes, though. I also just made a change to explicitly check the event type.

eric replied:
You do have to re-watch (because that sends the watch to the zookeeper server).

The register call is a library-specific construct — you only need to
do that part once.

ryanlecompte replied:
Ah! That makes sense. I'll ensure that the register part only happens once.

eric commented on lib/redis_failover/util.rb:
```diff
@@ -3,6 +3,8 @@ module RedisFailover
   module Util
     extend self
 
+    ZK_PATH = '/redis_failover_nodes'
```
It might make sense to allow this to be configurable — either the actual key name, or at least a prefix...

eric commented on lib/redis_failover/zk_client.rb:
```diff
@@ -0,0 +1,62 @@
+
+    def build_client
+      @lock.synchronize do
+        @client.reopen and return if @client
```
Personal preference: Lines like this scare me. It makes it too easy to make mistakes or have others in the future mess up the logic accidentally. I would much rather have this be a multi-line condition.

Also, the return will only happen if `reopen()` doesn't return `false` or `nil` — are those intentional cases?

ryanlecompte replied:
Good point! I normally don't write code this way, not sure what I was thinking. I've fixed this to be more explicit.

eric commented on lib/redis_failover/client.rb:
```diff
@@ -67,29 +66,25 @@ class Client
-
       Util.logger = options[:logger] if options[:logger]
+      @zkservers = options.fetch(:zkservers) { raise ArgumentError, ':zkservers required'}
+      @zkclient = ZkClient.new(@zkservers)
```
I'm trying to decide if it would make sense to allow for a zookeeper instance to be passed in to this class...

If I'm already using zookeeper in my processes, does it make sense to have twice the connections?

ryanlecompte replied:
Good point. Right now I have a custom ZkClient wrapper to handle session failures, etc. I guess I could support passing in an existing ZK instance, but at least if I have my own instance I can control setting my own custom watches, etc. I'll keep it in mind though.

eric commented:
I totally agree that calling `sync()` will be required to ensure that the clients all see the same thing.

To solve the issue @freels mentioned, I'm having trouble deciding if it would be solved simply by having an election process to allow a node manager that is having problems to be replaced by another, or if it would actually require some sort of voting (which is also something that can be orchestrated in zookeeper).

I would definitely be interested in comments people had in this regard.

eric commented:
A couple other random things I've been thinking about that always makes things interesting:

When forking with ruby (with passenger, resque, etc), the zookeeper library can act up (segfault) if it is not closed before forking, so it would be worthwhile to provide a way for users of the client to call a method to close and open the zookeeper session (or provide an accessor to assign a new zookeeper client instance).

ryanlecompte committed:
incorporate code review comments

ryanlecompte commented:
In my experience, the redis library itself also has problems when preforking servers like this. What I've done in the past is simply recreate my redis connection instance in an after_fork type hook that passenger provides. I think in this case, since RedisFailover::Client not only maintains a ZK client, but also multiple Redis clients, it probably makes sense to just recreate the entire RedisFailover::Client instance in an after-fork hook?

eric commented:
It totally does. We'll just need to make sure the redis client has a
close method that closes the ZK object as well as the redis ones.

ryanlecompte committed:
rely on watches instead of polling ZK for each redis operation

ryanlecompte committed:
add default znode path override option

ryanlecompte committed:
cosmetic

ryanlecompte commented:
I think I've addressed most of the issues raised in the pull request (apart from a graceful #close method for pre-forking servers, since right now there is a bug in #close! for the ZK gem). I'll merge this into master for now.
